lib/events/roomcanonicalaliasevent.h

```diff
@@ -0,0 +1,86 @@
+// Contains code for RoomCanonicalAlias Event
+#pragma once
+
+#include "stateevent.h"
+
+namespace Quotient {
+namespace EventContent{
+    template <typename T1, typename T2>
+    class SimpleDualContent {
+    public:
+        using first_value_type = T1;
+        using second_value_type = T2;
+
+        template <typename TT1, typename TT2>
+        SimpleDualContent(QString Key1Name, TT1&& value1, QString Key2Name,
+                          TT2&& value2)
+            : value1(std::forward<TT1>(value1))
+            , value2(std::forward<TT2>(value2))
+            , key1(std::move(Key1Name))
+            , key2(std::move(Key2Name))
+        { }
+
+        SimpleDualContent(const QJsonObject& json, QString key1Name,
+                          QString key2Name)
+            : value1(fromJson<T1>(json[key1Name]))
+            , value2(fromJson<T2>(json[key2Name]))
+            , key1(std::move(key1Name))
+            , key2(std::move(key2Name))
+        { }
+
+        QJsonObject toJson() const
+        {
+            return { { key1, Quotient::toJson(value1) },
+                     { key2, Quotient::toJson(value2) } };
+        }
+
+    public:
+        T1 value1;
+        T2 value2;
+
+    protected:
+        QString key1;
+        QString key2;
+    };
+} // namespace EventContent
+
+class RoomCanonicalAliasEvent
+    : public StateEvent<EventContent::SimpleDualContent<QString, QStringList>> {
+public:
+    DEFINE_EVENT_TYPEID("m.room.canonical_alias", RoomCanonicalAliasEvent)
+
+    explicit RoomCanonicalAliasEvent(const QJsonObject& obj)
+        : StateEvent(typeId(), obj, QStringLiteral("alias"),
+                     QStringLiteral("alt_aliases"))
+    { }
+
+    RoomCanonicalAliasEvent(const QString& server, const QString& alias,
+                            const QStringList& alt_aliases)
+        : StateEvent(typeId(), matrixTypeId(), server, QStringLiteral("alias"),
+                     alias, QStringLiteral("alt_aliases"), alt_aliases)
+    { }
+
+    // For compatibility used at Room::setCanonicalAlias
+    explicit RoomCanonicalAliasEvent(const QString& value1)
+        : RoomCanonicalAliasEvent(value1, QStringList())
+    { }
+
+    // Because, MSC2432 specifies, that alt_aliases may be present
+    // without aliases as well
+    explicit RoomCanonicalAliasEvent(const QStringList& value2)
+        : RoomCanonicalAliasEvent(QString(), value2)
+    { }
```

Comment:
Sir, in this document: https://docs.google.com/document/d/1NNDkobiFLeUkJtyj0H6qvKIedgvIkZnFKo78-03cGEk/edit

It is mentioned that: "alt_aliases" may exist without "alias"; this can mean, for example, that the main alias has been deleted.

Reply:
Right but imagine if you were a client author - would you expect `Room::setLocalAliases(QStringList)` to _delete_ the canonical alias, especially since it wasn't the behaviour of the library until this PR?

Reply:
Yes sir, so can we except an empty canonical alias in the constructor?

Reply:
I'm not sure I understand the question.

Reply:
Sir we will make a single constructor, `RoomCanonicalAliasEvent(QString canonicalAlias, QStringList altAliases)` should this constructor allow canonicalAlias to be an empty string?

Reply:
Ah - no I don't think so. Making an event object with an empty/missing canonical alias is very much a corner case, and it's worth to make it explicit in the code when really needed. And the spec says [there's no difference between an empty alias and no alias at all](https://matrix.org/docs/spec/client_server/latest#m-room-canonical-alias).

So I'd make `RoomCanonicalAliasEvent(QString canonicalAlias, QStringList altAliases = {})`, and anyone who really needs to clear away the canonical alias can easily (but explicitly, which is a good thing in this case) write `RoomCanonicalAliasEvent({}, aliases)` or (absent aliases) `RoomCanonicalAliasEvent({})`.